_source/_posts/2023-08-30-passwordless-authentication.md

```diff
@@ -0,0 +1,203 @@
+
+There are a couple of components at play. From the user's perspective, they interact with a website and a device like their phone. From a technology perspective, there's WebAuthn and CTAP2. I'll show screenshots of the user perspective underlying network calls and describe the hand-off between WebAuthn and CTAP2.
+
+## Registering a user for passkeys
+
+The user opens their favorite site in the browser. In the FIDO2 world, the site that makes the WebAuthn calls is the _Relying Party_, and the browser is the _Client_. If your site uses a single-page application frontend, the relying party includes both the frontend and backend server. 
+
+{% img blog/passwordless-authentication/webauthnio.jpg alt:"webauthn.io site showing a user name field an buttons to register or authenticate" width:"800" %}{: .center-image }
+
+The web app for the site displays a form to gather user information, an email address, or a user name. The user enters the required information and presses **Register**. The web app recognizes the browser as FIDO2 capable and asks if the user wants to sign in using a passkey instead of a password. Of course, the user says, "Yes!"
+
+{% img blog/passwordless-authentication/webauthnio-register.jpg alt:"webauthn.io site showing a user name entered in the form and a modal from the browser asking to create a passkey" width:"800" %}{: .center-image }
+
+The web app calls the server with credential request options, such as the username and registration-related data. You can open debugging capabilities in your browser and watch the network requests. Here, we see the payload for the call, which includes the username.
+
+{% img blog/passwordless-authentication/webauthnio-register-payload.jpg alt:"webauthn.io payload to register a user includes the username along with registration metadata configurable in advanced settings, such as encryption algorithm" width:"800" %}{: .center-image }
+
+The server responds with information about the relying party, a user ID, and a cryptographically random value for the registration process called a "challenge." 
+
+{% img blog/passwordless-authentication/webauthnio-register-response.jpg alt:"webauthn.io response includes relying party id, user name and id, and the challenge" width:"800" %}{: .center-image }
+
+The user selects a device for the passkey, like a phone or tablet. The web app uses WebAuthn to generate a new credential, sending the response information from the server, including the challenge. The credential call from the browser to the authenticator is possible via CTAP2.
+
+The authenticator takes over, prompting the user for consent and biometric information. The authenticator generates a public/private key set for the website. The authenticator signs the public key with an attestation certificate for the device hardware. Signing the key with the attestation certificate proves the keyset is from a specific device. The authenticator returns the public key in the WebAuthn credential response.
```

Comment:
are keypairs referred to as key sets in WebAuthn? I haven't heard pairs called sets before but I may be missing context

Reply:
Good point, "key sets" are often used within OAuth2 to refer to an Authorization Server's list of available public keys (e.g. the `/oauth2/v1/keys` endpoint).  I'd use "key pair" here.